arc-mlir/src/include/arc/arc-dialect.h

```diff
@@ -23,21 +23,94 @@
 #ifndef ARC_DIALECT_H_
 #define ARC_DIALECT_H_
 
+#include "mlir/IR/StandardTypes.h"
+#include <mlir/IR/Builders.h>
 #include <mlir/IR/Dialect.h>
 #include <mlir/IR/Function.h>
-#include <mlir/IR/Builders.h>
 #include <mlir/IR/Operation.h>
 
 using namespace mlir;
 
 namespace arc {
 
+namespace ArcTypes {
+
+enum Kind {
+  // These kinds will be used by Arc.
+  Appender = Type::Kind::FIRST_PRIVATE_EXPERIMENTAL_0_TYPE,
+};
+
+bool isValueType(Type type);
+
+bool isBuilderType(Type type);
+
+struct AppenderTypeStorage : public TypeStorage {
+  AppenderTypeStorage(Type mergeType) : mergeType(mergeType) {}
+
+  Type mergeType;
+
+  using KeyTy = Type;
+
+  bool operator==(const KeyTy &key) const { return key == KeyTy(mergeType); }
+
+  static llvm::hash_code hashKey(const KeyTy &key) {
+    return llvm::hash_combine(key);
+  }
+
+  static KeyTy getKey(Type mergeType) { return KeyTy(mergeType); }
+
+  static AppenderTypeStorage *construct(TypeStorageAllocator &allocator,
+                                        const KeyTy &key) {
+    return new (allocator.allocate<AppenderTypeStorage>())
+        AppenderTypeStorage(key);
+  }
+};
+
+class AppenderType
+    : public Type::TypeBase<AppenderType, Type, AppenderTypeStorage> {
+public:
+  using Base::Base;
+
+  static bool kindof(ArcTypes::Kind kind) { return kind == ArcTypes::Appender; }
+
+  static AppenderType get(Type mergeType) {
+    mlir::MLIRContext *context = mergeType.getContext();
+    return Base::get(context, ArcTypes::Appender, mergeType);
+  }
+
+  static AppenderType getChecked(Type mergeType, Location location) {
+    mlir::MLIRContext *context = mergeType.getContext();
+    return Base::getChecked(location, context, ArcTypes::Appender, mergeType);
+  }
+
+  static LogicalResult
+  verifyConstructionInvariants(llvm::Optional<Location> loc,
+                               MLIRContext *context, Type type) {
+    //         if (!type.isa<IntegerType>()) {
+    //           if (loc)
+    //             context->emitError(loc) << "non integer-type passed to
+    //             'AppenderType'";
+    //           return failure();
+    //         }
+    return success();
+  }
+
+  Type getMergeType() { return getImpl()->mergeType; }
+};
```

Comment:
This should probably be moved into a `arc-dialect.cpp` but I'm not sure which parts to transfer. How does C++ handle forward declarations compared to C?

Reply:
```patch
diff --git a/arc-mlir/src/arc-dialect.cpp b/arc-mlir/src/arc-dialect.cpp
index cd8f967..bef6331 100644
--- a/arc-mlir/src/arc-dialect.cpp
+++ b/arc-mlir/src/arc-dialect.cpp
@@ -110,6 +110,8 @@ Type ArcDialect::parseType(DialectAsmParser &parser) const {
   return nullptr;
 }
 
+Type ArcTypes::AppenderType::getMergeType() { return getImpl()->mergeType; };
+
 //===----------------------------------------------------------------------===//
 // Arc Type Printer
 //===----------------------------------------------------------------------===//
diff --git a/arc-mlir/src/include/arc/arc-dialect.h b/arc-mlir/src/include/arc/arc-dialect.h
index d90ffec..b98395c 100644
--- a/arc-mlir/src/include/arc/arc-dialect.h
+++ b/arc-mlir/src/include/arc/arc-dialect.h
@@ -95,7 +95,7 @@ public:
     return success();
   }
 
-  Type getMergeType() { return getImpl()->mergeType; }
+  Type getMergeType();
 };
 } // namespace ArcTypes
```
is how to move an inline declaration to the module.

Reply:
Oh, then it makes sense, I will do it for all the methods.

Reply:
For simple things like that, which fits on one line, I'm fine with doing it inline.